### PR TITLE
Release-v2.0.0

### PR DESCRIPTION
**Pull request by rschouten97**
Changed to use of Access Token Authentication instead of Certificate Based Authentication
Removed building an additional PS session to hold the EXO session
Added check if all mandatory values from the configuration object are available
Added check if the correlation fields are available
Added check if all mandatory values from the account object are available
Updated logging to latest best practices

### CODE REVIEW

**Review by rhouthuijzen — COMMENTED** [review]
An App Registration in Azure AD is required. Please follow the Microsoft documentation as reference to configure the App Registration correctly

The link is based on a certificate

**Review by rhouthuijzen — COMMENTED** [review]
Microsoft Exchange Online PowerShell V2 module 

Change to V3.1 (multiple times in readme)

File: create.correlateUpdateMailboxFolderPermission.ps1 [hunk 1]
```diff
@@ -59,500 +67,264 @@ function Resolve-HTTPError {
...
-            if ([String]::IsNullOrEmpty($auditErrorMessage)) {
-                $auditErrorMessage = $ex.Exception.Message
-            }
-        
-            Write-Verbose "Error at Line '$($ex.InvocationInfo.ScriptLineNumber)': $($ex.InvocationInfo.Line). Error: $($verboseErrorMessage)"
+    try {
+        Write-Verbose "Updating folder permissions for mailbox [$($aRef.userPrincipalName) ($($aRef.Guid))]: $($mailboxSplatParams | ConvertTo-Json)"
+
+        # Get Mailbox "Calendar" folder name
+        $mailboxFolderId = Get-MailboxFolderStatistics -Identity $mailbox.Guid -FolderScope Calendar | Where-Object { $_.FolderType -eq 'Calendar' } | Select-Object Name
```
Comment (rhouthuijzen) [inline create.correlateUpdateMailboxFolderPermission.ps1:293]: I get this result: @{Name=Agenda}

Suggestion:
(Get-MailboxFolderStatistics -Identity $mailbox.Guid -FolderScope Calendar | Where-Object { $_.FolderType -eq 'Calendar' } | Select-Object Name).Name

Result: Agenda

File: create.correlateUpdateMailboxRegionalConfigurationAndFolderPermission.ps1 [hunk 2]
```diff
@@ -88,630 +74,303 @@ function Resolve-HTTPError {
...
+                IsError = $True
+            })
+    }
+
+    # Update Mailbox Folder Permission
+    try {
+        Write-Verbose "Updating folder permissions for mailbox [$($aRef.userPrincipalName) ($($aRef.Guid))]: $($mailboxSplatParams | ConvertTo-Json)"
+
+        # Get Mailbox "Calendar" folder name
+        $mailboxFolderId = Get-MailboxFolderStatistics -Identity $mailbox.Guid -FolderScope Calendar | Where-Object { $_.FolderType -eq 'Calendar' } | Select-Object Name
```
Comment (rhouthuijzen) [inline create.correlateUpdateMailboxRegionalConfigurationAndFolderPermission.ps1:300]: I get this result: @{Name=Agenda}

Suggestion:
(Get-MailboxFolderStatistics -Identity $mailbox.Guid -FolderScope Calendar | Where-Object { $_.FolderType -eq 'Calendar' } | Select-Object Name).Name

Result: Agenda

File: delete.UpdateMailboxAutoReplyConfiguration.ps1 [hunk 3]
```diff
@@ -78,498 +68,243 @@ function Resolve-HTTPError {
...
-                    }
-                    $returnobject.Keys | ForEach-Object { Remove-Variable $_ -ErrorAction SilentlyContinue }
-                    Write-Output $returnobject 
-                }
-            }
+    try {
+        Write-Verbose "Updating autoreply configuration for mailbox [$($aRef.userPrincipalName) ($($aRef.Guid))]: $($mailboxSplatParams | ConvertTo-Json)"
+
+        # Get Mailbox "Calendar" folder name
+        $mailboxFolderId = Get-MailboxFolderStatistics -Identity $mailbox.Guid -FolderScope Calendar | Where-Object { $_.FolderType -eq 'Calendar' } | Select-Object Name
```
Comment (rhouthuijzen) [inline delete.UpdateMailboxAutoReplyConfiguration.ps1:273]: $mailboxFolderId is not used. So I think the Get-MailboxFolderStatistics can be removed

File: revoke.sharedmailboxes.ps1 [hunk 4]
```diff
@@ -69,576 +55,331 @@ function Resolve-HTTPError {
...
+                    $FullAccessPermissionSplatParams = @{
+                        Identity        = $pRef.id
+                        User            = $aRef.Guid
+                        AccessRights    = 'FullAccess'
+                        InheritanceType = 'All'
+                        AutoMapping     = $AutoMapping
+                    } 
+
+                    if ($dryRun -eq $false) {
+                        $addFullAccessPermission = Add-MailboxPermission @FullAccessPermissionSplatParams -ErrorAction Stop
```
Comment (rhouthuijzen) [inline revoke.sharedmailboxes.ps1:225]: Add-MailboxPermission --> Remove-MailboxPermission

File: revoke.sharedmailboxes.ps1 [hunk 5]
```diff
@@ -69,576 +55,331 @@ function Resolve-HTTPError {
...
-                        $module = Import-Module $ModuleName -Cmdlet $commands
-                        [Void]$verboseLogs.Add("Imported module $ModuleName")
+    # revoke Exchange Online Mailbox permission
+    foreach ($permission in $pRef.Permissions) {
+        switch ($permission) {
+            "Full Access" {
+                try {
+                    Write-Verbose "Revoking permission [FullAccess] to mailbox [$($pRef.Name) ($($pRef.id))] for user [$($aRef.UserPrincipalName) ($($aRef.Guid))]"
+
+                    $FullAccessPermissionSplatParams = @{
```
Comment (rhouthuijzen) [inline revoke.sharedmailboxes.ps1:216]: Update SplatParms to:

                    Identity        = $pRef.id
                    User            = $aRef.Guid
                    AccessRights    = 'FullAccess'
                    InheritanceType = 'All'
                    Confirm         = $false

File: revoke.sharedmailboxes.ps1 [hunk 6]
```diff
@@ -69,576 +55,331 @@ function Resolve-HTTPError {
...
-                [Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls -bor [Net.SecurityProtocolType]::Tls11 -bor [Net.SecurityProtocolType]::Tls12
+        # Create access token
+        Write-Verbose "Creating Access Token"
 
-                $auditLogs = [System.Collections.Generic.List[PSCustomObject]]::new()
+        $baseUri = "https://login.microsoftonline.com/"
+        $authUri = $baseUri + "$AADTenantId/oauth2/token"
+        
+        $body = @{
+            revoke_type   = "client_credentials"
```
Comment (rhouthuijzen) [inline revoke.sharedmailboxes.ps1:168]: revoke_type --> grant_type

File: dynamicPermission.SharedMailboxes.ReferenceExample.ps1 [hunk 7]
```diff
@@ -0,0 +1,735 @@
...
+                $FullAccessPermissionSplatParams = @{
+                    Identity        = $permission.Name
+                    User            = $aRef.Guid
+                    AccessRights    = 'FullAccess'
+                    InheritanceType = 'All'
+                    AutoMapping     = $AutoMapping
+                } 
+
+                if ($dryRun -eq $false) {
+                    $addFullAccessPermission = Add-MailboxPermission @FullAccessPermissionSplatParams -ErrorAction Stop
```
Comment (rhouthuijzen) [inline dynamicPermission.SharedMailboxes.ReferenceExample.ps1:432]: Not tested. But Add-MailboxPermissions seems odd when revoking permissions

File: dynamicPermission.SharedMailboxes.ReferenceExample.ps1 [hunk 8]
```diff
@@ -0,0 +1,735 @@
...
+
+                $sendAsPermissionSplatParams = @{
+                    Identity     = $permission.Name
+                    Trustee      = $aRef.Guid
+                    AccessRights = 'SendAs'
+                    Confirm      = $false
+                } 
+
+                if ($dryRun -eq $false) {
+                    $addSendAsPermission = Add-RecipientPermission @sendAsPermissionSplatParams -ErrorAction Stop
```
Comment (rhouthuijzen) [inline dynamicPermission.SharedMailboxes.ReferenceExample.ps1:486]: Not tested. But Add-RecipientPermission seems odd when revoking permissions

File: revoke.sharedmailboxes.ps1 [hunk 9]
```diff
@@ -21,39 +17,29 @@ $auditLogs = [System.Collections.Generic.List[PSCustomObject]]::new()
...
+    $true { $VerbosePreference = "Continue" }
+    $false { $VerbosePreference = "SilentlyContinue" }
 }
 $InformationPreference = "Continue"
 $WarningPreference = "Continue"
 
+# Define configuration properties as required
+$requiredConfigurationFields = @("AzureADOrganization", "AzureADTenantId", "AzureADAppId", "AzureADAppSecret")
+
 # Used to connect to Exchange Online in an unattended scripting scenario using a certificate.
```
Comment (rhouthuijzen) [inline revoke.sharedmailboxes.ps1:29]: Certifcate is nolonger used

File: revoke.groups.ps1 [hunk 10]
```diff
@@ -21,39 +17,28 @@ $auditLogs = [System.Collections.Generic.List[PSCustomObject]]::new()
...
+    $true { $VerbosePreference = "Continue" }
+    $false { $VerbosePreference = "SilentlyContinue" }
 }
 $InformationPreference = "Continue"
 $WarningPreference = "Continue"
 
+# Define configuration properties as required
+$requiredConfigurationFields = @("AzureADOrganization", "AzureADTenantId", "AzureADAppId", "AzureADAppSecret")
+
 # Used to connect to Exchange Online in an unattended scripting scenario using a certificate.
```
Comment (rhouthuijzen) [inline revoke.groups.ps1:29]: Certifcate is nolonger used

**Comment by rschouten97** [comment]
> An App Registration in Azure AD is required. Please follow the Microsoft documentation as reference to configure the App Registration correctly
> 
> The link is based on a certificate



> An App Registration in Azure AD is required. Please follow the Microsoft documentation as reference to configure the App Registration correctly
> 
> The link is based on a certificate

The official Microsoft Documentation still only references the use of Certificate Based Authentication.
I removed the notes that you should follow the microsoft docs and added notes that our setup differs by using the access token.

**Comment by rschouten97** [comment]
> Microsoft Exchange Online PowerShell V2 module
> 
> Change to V3.1 (multiple times in readme)

Changed all references from v2 to v3.1. However, the microsoft docs articles still have the v2 url, I left this unchanged of course.

File: delete.UpdateMailboxAutoReplyConfiguration.ps1 (hunk 3 above)
Comment (rschouten97) [reply delete.UpdateMailboxAutoReplyConfiguration.ps1:273]: Removed

File: revoke.sharedmailboxes.ps1 (hunk 5 above)
Comment (rschouten97) [reply revoke.sharedmailboxes.ps1:216]: Updated

File: revoke.sharedmailboxes.ps1 (hunk 4 above)
Comment (rschouten97) [reply revoke.sharedmailboxes.ps1:225]: Corrected

File: revoke.groups.ps1 (hunk 10 above)
Comment (rschouten97) [reply revoke.groups.ps1:29]: corrected.

File: revoke.sharedmailboxes.ps1 (hunk 9 above)
Comment (rschouten97) [reply revoke.sharedmailboxes.ps1:29]: corrected.

File: create.correlateUpdateMailboxFolderPermission.ps1 (hunk 1 above)
Comment (rschouten97) [reply create.correlateUpdateMailboxFolderPermission.ps1:291]: Good catch, corrected.

File: create.correlateUpdateMailboxRegionalConfigurationAndFolderPermission.ps1 (hunk 2 above)
Comment (rschouten97) [reply create.correlateUpdateMailboxRegionalConfigurationAndFolderPermission.ps1:298]: Good catch. corrected

File: dynamicPermission.SharedMailboxes.ReferenceExample.ps1 (hunk 7 above)
Comment (rschouten97) [reply dynamicPermission.SharedMailboxes.ReferenceExample.ps1]: corrected

File: revoke.sharedmailboxes.ps1 [hunk 11]
```diff
@@ -69,576 +54,330 @@ function Resolve-HTTPError {
...
-                [Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls -bor [Net.SecurityProtocolType]::Tls11 -bor [Net.SecurityProtocolType]::Tls12
+        # Create access token
+        Write-Verbose "Creating Access Token"
 
-                $auditLogs = [System.Collections.Generic.List[PSCustomObject]]::new()
+        $baseUri = "https://login.microsoftonline.com/"
+        $authUri = $baseUri + "$AADTenantId/oauth2/token"
+        
+        $body = @{
+            revoke_type   = "client_credentials"
```
Comment (rhouthuijzen) [inline revoke.sharedmailboxes.ps1:167]: This should be grant_type

**Comment by rhouthuijzen** [comment]
Possibly change Azure AD role:

Assign Azure AD roles to the application
Azure AD has more than 50 admin roles available. The Global Administrator

The Global Administrator Rol is not required for the following ExO functions (tested):

    "Get-EXOMailbox"
    "Set-MailboxFolderPermission"
    "Get-MailboxFolderStatistics"
    "Add-MailboxFolderPermission"
    "Remove-MailboxFolderPermission"
    "Remove-MailboxPermission"
    "Remove-RecipientPermission"
    "Get-User"
    "Add-MailboxPermission"
    "Add-RecipientPermission"
    "Set-Mailbox"

File: revoke.sharedmailboxes.ps1 (hunk 11 above)
Comment (rschouten97) [reply revoke.sharedmailboxes.ps1:167]: Corrected

**Review by rhouthuijzen — APPROVED** [review]
Approved. Used multiple times in implementations.

**Pull request merged**